vita3k/shader/src/spirv_recompiler.cpp

```diff
@@ -1029,13 +1030,44 @@ static spv::Function *make_vert_finalize_function(spv::Builder &b, const SpirvSh
     return vert_fin_func;
 }
 
+static SpirvCode optimize_shader_spirv(const SpirvCode &input) {
+    SpirvCode result;
+
+    // target should change depending on backend, but everything goes through SPIRV-Cross now so its not a big deal now
+    const spv_target_env target = SPV_ENV_OPENGL_4_1;
+
+    // need to use c api to get diagnostic information
+    spv_context validator = spvContextCreate(target);
+    spv_diagnostic diagnostic{};
+    spv_result_t validator_result = spvValidateBinary(validator, input.data(), input.size(), &diagnostic);
+    if (validator_result != SPV_SUCCESS) {
+        LOG_WARN("Validator failed. {} (pos: {})",
+            diagnostic->error, diagnostic->position.index);
+    }
+    spvDiagnosticDestroy(diagnostic);
+    spvContextDestroy(validator);
+
+    spvtools::OptimizerOptions optimizer_options;
+    optimizer_options.set_run_validator(false);
+
+    spvtools::Optimizer optimizer(target);
+
+    optimizer.RegisterPerformancePasses();
+    optimizer.Run(input.data(), input.size(), &result, optimizer_options);
```

Comment:
This one returns error! You can return the raw input if error occur to make it not crash.